frame/name-service/src/lib.rs

```diff
@@ -0,0 +1,487 @@
+				ensure!(r.owner == sender, Error::<T>::NotRegistrationOwner);
+				ensure!(r.expiry > block_number, Error::<T>::RegistrationExpired);
+
+				r.owner = to.clone();
+
+				// TODO: transfer deposit over?
+
+				Self::deposit_event(Event::<T>::Transfer { from: sender, to });
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(0)]
+		pub fn renew(origin: OriginFor<T>, name_hash: NameHash, periods: u32) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			Registrations::<T>::try_mutate(name_hash, |maybe_registration| {
+				let r = maybe_registration.as_mut().ok_or(Error::<T>::RegistrationNotFound)?;
+
+				let fee = Self::length_fee(periods);
+
+				// withdraw fees from account
+				let imbalance = T::Currency::withdraw(
+					&sender,
+					fee,
+					WithdrawReasons::FEE,
+					ExistenceRequirement::KeepAlive,
+				)?;
+
+				let block_number = frame_system::Pallet::<T>::block_number();
+
+				let expiry_new = match r.expiry > block_number {
+					true => r.expiry.saturating_add(Self::length(periods)),
+					false => block_number.saturating_add(Self::length(periods)),
+				};
+
+				r.expiry = expiry_new;
+
+				T::RegistrationFeeHandler::on_unbalanced(imbalance);
+
+				Self::deposit_event(Event::<T>::NameRenewed { name_hash, expires: expiry_new });
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(0)]
+		pub fn set_address(
+			origin: OriginFor<T>,
+			name_hash: NameHash,
+			address: T::AccountId,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			let registration =
+				Registrations::<T>::get(name_hash).ok_or(Error::<T>::RegistrationNotFound)?;
+			ensure!(registration.owner == sender, Error::<T>::NotRegistrationOwner);
+
+			ensure!(
+				registration.expiry > frame_system::Pallet::<T>::block_number(),
+				Error::<T>::RegistrationExpired
+			);
+
+			Resolvers::<T>::insert(name_hash, address.clone());
+
+			Self::deposit_event(Event::<T>::AddressSet { name_hash, address });
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn deregister(origin: OriginFor<T>, name_hash: NameHash) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			Self::do_deregister(name_hash, Some(sender))?;
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn force_register(
+			origin: OriginFor<T>,
+			name_hash: NameHash,
+			who: T::AccountId,
+			periods: u32,
+		) -> DispatchResult {
+			T::RegistrationManager::ensure_origin(origin)?;
+			let deposit: BalanceOf<T> = Zero::zero();
+			Self::do_register(name_hash, who, deposit, periods)?;
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn force_deregister(origin: OriginFor<T>, name_hash: NameHash) -> DispatchResult {
+			T::RegistrationManager::ensure_origin(origin)?;
+			Self::do_deregister(name_hash, None)?;
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn remove_commitment(
+			origin: OriginFor<T>,
+			commitment_hash: CommitmentHash,
+		) -> DispatchResult {
+			ensure_signed_or_root(origin)?;
```

Comment:
Why the root here?

Reply:
because it is a permissionless function, and so we can allow root, for example governance to call this function.

question should be "why not"?